src/core/small_string.cc

```diff
@@ -54,55 +54,36 @@ size_t SmallString::UsedThreadLocal() {
 
 static_assert(sizeof(SmallString) == 16);
 
-// we should use only for sizes greater than kPrefLen
 size_t SmallString::Assign(std::string_view s) {
   DCHECK_GT(s.size(), kPrefLen);
-
+  DCHECK(CanAllocate(s.size()));
   uint8_t* realptr = nullptr;
 
-  if (size_ == 0) {
-    // packed structs can not be tied here.
+  // If the allocation is large enough and space efficient, we can avoid allocating
+  if (s.size() >= size_ || s.size() * 2 < MallocUsed()) {
```

Comment:
something is wrong with this condition, imho

Reply:
Ah, it's just the inverse of the comment, I'll update it